### PR TITLE
Feature/traffic usage

### PR DESCRIPTION
**Pull request by daizong**
support traffic usage for proxy and super proxy

### CODE REVIEW

File: usage/usage.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+import "sync/atomic"
+
+type ProxyUsage struct {
+	Incoming uint64 //byte size
+	Outgoing uint64 //byte size
+}
+
+func (u *ProxyUsage) AddIncomingSize(n uint64) {
+	atomic.AddUint64(&u.Incoming, n)
```
Comment (cszichao) [inline usage/usage.go]: 是否会成为瓶颈，找翔宇测一下，同时开上百个线程测一下，如果影响的话是否需要开新goroutine？

**Comment by daizong** [comment]
我改为每个usage单用一个goroutine来统计，通过chan来传递数据，chan的大小默认是1000，写chan的操作考虑开销没有再开goroutine来处理。

File: proxy/http/http.go [hunk 2]
```diff
@@ -55,6 +58,7 @@ func (r *Request) Reset() {
 	r.isTLS = false
 	r.tlsServerName = ""
 	r.hostWithPort = ""
+	r.readSize = 0
```
Comment (cszichao) [inline proxy/http/http.go]: not zero here, add request line size.

File: proxy/http/http.go [hunk 3]
```diff
@@ -233,11 +259,13 @@ func (r *Response) ReadFrom(discardBody bool, reader *bufio.Reader) error {
 	if err := util.WriteWithValidation(r.writer, respLineBytes); err != nil {
 		return util.ErrWrapper(err, "fail to write start line of response")
 	}
+	r.writeSize += len(respLineBytes)
 
 	//read & write the headers
 	var hijackerBodyWriter io.Writer
 	if err := copyHeader(&r.header, reader, r.writer,
 		func(rawHeader []byte) {
+			r.writeSize += len(rawHeader)
```
Comment (cszichao) [inline proxy/http/http.go]: not exactly the returned header traffic, cause the header has been translated and modified

File: proxy/http/http.go [hunk 4]
```diff
@@ -135,6 +139,7 @@ func (r *Request) WriteHeaderTo(writer *bufio.Writer) error {
 	//read & write the headers
 	return copyHeader(&r.header, r.reader, writer,
 		func(rawHeader []byte) {
+			r.readSize += len(rawHeader)
```
Comment (cszichao) [inline proxy/http/http.go]: not exactly the incoming header traffic, cause the header has been translated and modified

**Pull request closed without merging**